test/WebJobs.Extensions.Tests/Extensions/Http/ClaimsPrincipalEndToEndTests.cs

```diff
@@ -0,0 +1,206 @@
+            string easyAuthHeaderValue = Convert.ToBase64String(Encoding.UTF8.GetBytes(UserIdentityJson));
+            request.Headers.Add("x-ms-client-principal", easyAuthHeaderValue);
+
+            var method = typeof(TestFunctions).GetMethod(nameof(TestFunctions.HttpRequest));
+            _host.Call(method, new { req = request });
+
+            Assert.Equal(Name, globalPrincipal.Identity.Name);
+            Assert.Equal(UserAuthType, globalPrincipal.Identity.AuthenticationType);
+            Claim testClaim = globalPrincipal.Claims.Where(claim => string.Equals(claim.Type, UserClaimType)).First();
+            Assert.Equal(testClaim.Value, UserClaimValue);
+        }
+
+        [Fact]
+        public void ClaimsPrincipalBindingWithFunctionsKey()
+        {
+            var request = HttpTestHelpers.CreateHttpRequest("GET", "http://functions.com/api/TestIdentityBindings");
+            string keyHeaderValue = Convert.ToBase64String(Encoding.UTF8.GetBytes(KeyIdentityJson));
+            request.Headers.Add("x-ms-functions-key-identity", keyHeaderValue);
+
+            var method = typeof(TestFunctions).GetMethod(nameof(TestFunctions.HttpRequest));
+            _host.Call(method, new { req = request });
+
+            Assert.Equal(KeyAuthType, globalPrincipal.Identity.AuthenticationType);
+            Claim idClaim = globalPrincipal.Claims.Where(claim => string.Equals(claim.Type, KeyIdClaimType)).First();
+            Assert.Equal(idClaim.Value, KeyIdClaimValue);
+            Claim levelClaim = globalPrincipal.Claims.Where(claim => string.Equals(claim.Type, KeyLevelClaimType)).First();
+            Assert.Equal(levelClaim.Value, KeyLevelClaimValue);
+        }
+
+        [Fact]
+        public void CLaimsPrincipalBindingWithBothEasyAuthAndFunctionsKey()
+        {
+            var request = HttpTestHelpers.CreateHttpRequest("GET", "http://functions.com/api/TestIdentityBindings");
+            string easyAuthHeaderValue = Convert.ToBase64String(Encoding.UTF8.GetBytes(UserIdentityJson));
+            request.Headers.Add("x-ms-client-principal", easyAuthHeaderValue);
+            string keyHeaderValue = Convert.ToBase64String(Encoding.UTF8.GetBytes(KeyIdentityJson));
+            request.Headers.Add("x-ms-functions-key-identity", keyHeaderValue);
+
+            var method = typeof(TestFunctions).GetMethod(nameof(TestFunctions.HttpRequest));
+            _host.Call(method, new { req = request });
+
+            Assert.Equal(Name, globalPrincipal.Identity.Name);
+            Assert.Equal(2, globalPrincipal.Identities.Count());
+        }
+
+        [Fact]
+        public void ClaimsPrincipalBindingWithNoIdentities()
+        {
+            var request = HttpTestHelpers.CreateHttpRequest("GET", "http://functions.com/api/TestIdentityBindings");
+            var method = typeof(TestFunctions).GetMethod(nameof(TestFunctions.HttpRequest));
+            _host.Call(method, new { req = request });
+
+            Assert.Null(globalPrincipal.Identity);
```

Comment:
>Assert.Null(globalPrincipal.Identity); [](start = 12, length = 38)

This seems inconsistent with how ASP.NET works. In ASP.NET, failing to authenticate results in a principal with `Identity.IsAuthenticated == false`. I suggest we do the same thing with the claims principal binding for consistency.

Reply:
Looking at [https://msdn.microsoft.com/en-us/library/system.security.claims.claimsprincipal.identity(v=vs.110).aspx](https://msdn.microsoft.com/en-us/library/system.security.claims.claimsprincipal.identity(v=vs.110).aspx) under Remarks, it looks like if no identities are provided, Identity is null. I don't know what identity we would set if it fails to find both an EasyAuth and a Functions Key identity.

---
In reply to: [181945812](https://github.com/Azure/azure-webjobs-sdk-extensions/pull/404#discussion_r181945812) [](ancestors = 181945812)

Reply:
My recollection is that they create add a `GenericIdentity` and set IsAuthenticated to `false`.

---
In reply to: [182183202](https://github.com/Azure/azure-webjobs-sdk-extensions/pull/404#discussion_r182183202) [](ancestors = 182183202,181945812)